modules/core/src/main/java/org/apache/ignite/internal/processors/rest/handlers/cluster/GridBaselineCommandHandler.java

```diff
@@ -0,0 +1,201 @@
+
+        Collection<? extends BaselineNode> srvrs = cluster.forServers().nodes();
+
+        return new GridBaselineCommandResponse(cluster.active(), cluster.topologyVersion(), currentBaseLine(), srvrs);
+    }
+
+    /**
+     * Set new baseline.
+     *
+     * @param baselineTop Initial collection of nodes.
+     * @param consistentIds Collection of consistent IDs to set.
+     */
+    private void set(Set<BaselineNode> baselineTop, List<Object> consistentIds) {
+        IgniteClusterEx cluster = ctx.grid().cluster();
+        
+        Map<Object, BaselineNode> srvrs = cluster.forServers().nodes().stream()
+            .collect(toMap(n -> String.valueOf(n.consistentId()), identity()));
+
+        for (Object consistentId : consistentIds) {
+            BaselineNode node = srvrs.get(consistentId);
+
+            if (node == null)
+                throw new IllegalStateException("Node not found for consistent ID: " + consistentId);
+
+            baselineTop.add(node);
+        }
+
+        cluster.setBaselineTopology(baselineTop);
+    }
+
+    /** {@inheritDoc} */
+    private GridBaselineCommandResponse handleAsyncUnsafe(GridRestRequest req) throws IgniteCheckedException {
+        assert req instanceof GridRestBaselineRequest : "Invalid command for baseline handler: " + req;
+
+        assert SUPPORTED_COMMANDS.contains(req.command());
+
+        GridRestBaselineRequest req0 = (GridRestBaselineRequest)req;
+
+        IgniteClusterEx cluster = ctx.grid().cluster();
+
+        List<Object> consistentIds = req0.consistentIds();
+
+        switch (req0.command()) {
+            case BASELINE_CURRENT_STATE: {
+                // No-op.
+
+                break;
+            }
+
+            case BASELINE_SET: {
+                Long topVer = req0.topologyVersion();
+
+                if (topVer == null && consistentIds == null)
+                    throw new IgniteCheckedException("Failed to handle request (either topVer or consistentIds should be specified).");
+
+                if (topVer != null)
+                    cluster.setBaselineTopology(topVer);
+                else
+                    set(new HashSet<>(), consistentIds);
+
+                break;
+            }
+
+            case BASELINE_ADD: {
+                if (consistentIds == null)
+                    throw new IgniteCheckedException(missingParameter("consistentIds"));
+
+                set(currentBaseLine(), consistentIds);
+
+                break;
+            }
+
+            case BASELINE_REMOVE: {
+                if (consistentIds == null)
+                    throw new IgniteCheckedException(missingParameter("consistentIds"));
+
+                Map<Object, BaselineNode> baseline = currentBaseLine().stream()
+                    .collect(toMap(n -> String.valueOf(n.consistentId()), identity()));
+
+                for (Object consistentId : consistentIds) {
+                    BaselineNode node = baseline.remove(consistentId);
```

Comment:
Another weird thing is that you use "String.valueOf(n.consistentId())" to create map but use "Object consistentId" to search in that map. Doesn't look consistent even if it works fine.